source/_posts/2016-09-30-spark-streaming-on-yarn.markdown

```diff
@@ -0,0 +1,324 @@
+
+Because Spark driver and Application Master share a single JVM, any error in Spark driver stops our long-running job.
+Fortunately it is possible to configure maximum number of attempts that will be made to re-run the application.
+It is reasonable to set higher value than default 2 (derived from YARN cluster property ```yarn.resourcemanager.am.max-attempts```).
+For me 4 works quite well, higher value may cause unnecessary restarts even if the reason of the failure is permanent.
+
+    spark-submit --master yarn --deploy-mode cluster \
+        --conf spark.yarn.maxAppAttempts=4
+
+If the application runs for days or weeks without restart or redeployment on highly utilized cluster, 4 attempts could be exhausted in few hours. 
+To avoid this situation, the attempt counter should be reset on every hour of so.
+
+    spark-submit --master yarn --deploy-mode cluster \
+        --conf spark.yarn.maxAppAttempts=4 \
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h
+ 
+Another important setting is a maximum number of executor failures before the application fails. 
+By default it is ```max(2 * num executors, 3)```, well suited for batch jobs but not for long-running jobs.
+The property comes with corresponding validity interval which also should be set.
+
+    spark-submit --master yarn --deploy-mode cluster \
+        --conf spark.yarn.maxAppAttempts=4 \
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h
+        
+For long-running jobs you could also consider to boost maximum number of task failures before giving up the job.
+By default tasks will be retried 4 times and then job fails.
+
+    spark-submit --master yarn --deploy-mode cluster \
+        --conf spark.yarn.maxAppAttempts=4 \
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8
+ 
+## Performance
+
+When Spark Streaming application is submitted to the cluster, YARN queue where the job runs must be defined.
+I strongly recommend to use YARN Capacity Scheduler and separate queue for long-running jobs.
+Without a separate YARN queue your long-running job will be preempted by a massive Hive query sooner or later.
+
+    spark-submit --master yarn --deploy-mode cluster \
+        --conf spark.yarn.maxAppAttempts=4 \
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue
+
+
+Another important performance factor for Spark Streaming job is processing time predictability. 
```

Comment:
Another important issue for Spark Streaming job is keeping processing time(micro batch duration) stable and highly predictable